---
title: "QC Analysis of ICB_Ravi mae"
author: "Nasim Bondar Sahebi"
date: "2023-11-07"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


# Install and load libraries.

```{r libraries}
library(ggplot2)
library(dplyr)
library(MultiAssayExperiment) 
library(limma)
library(EnhancedVolcano)
library(DT)
library(RColorBrewer)
library(readr)

```


# Data Loading and Preparation

## Load Gene Expression Data, 

**Data Overview**:

  - **Study Reference**: [PubMed ID 37024582](https://pubmed.ncbi.nlm.nih.gov/37024582/)
  - **Patient Count (RNA)**: 148
  - **Treatment Details**:
  - **PD-1/PD-L1**: 130 patients
  - **IO+chemo**: 2 patients
  - **IO+combo**: 16 patients

Load multiassay .rds file, extract clinical, expression data, and annotations; prepare gene expression data for analysis.

```{r load-expression-data}
# Load your multiassay result and extract clinical data , expression data and annotation

#load mae obj
mae <- readRDS("~/BHK lab/Ravi_Testing/data/ICB_Ravi.rds")

#extract Clinical data 
clin <- data.frame(colData(mae))

#extract the expression data
expr <- assays(mae)[["expr"]]

#extracting the annotation 
annot <- data.frame(rowData(mae@ExperimentList$expr))

# Display first few rows of the dataset
DT::datatable(expr[1:8, 1:4])
```

# Principal Component Analysis (PCA)

## Objective
Perform PCA on gene expression data to assess variation across different research centers. 

## Data Preparation

Following the 'RNA-seq differential expression analysis' section from the paper, steps for PCA-ready gene expression data include:

1. **Restrict expression data to protein-coding genes.**
2. **Convert expression data**: Convert from `log2(TPM+0.001)` to `log2(TPM+1)` for consistency.
3. **Remove low/zero expression genes**: Keep genes with expression at `log2TPM â‰¥ 0.5` across at least 30% of samples.


```{r protein-coing genes}
# Step 1: Restrict to Protein-Coding Genes
# Filter out non-protein-coding transcripts for focused analysis
annot_proteincoding <- annot[annot$gene_type == "protein_coding",] # 19158 protein coding genes.
expr <- expr[rownames(expr) %in% rownames(annot_proteincoding),]

# Step 2: Normalize Expression Data
# Convert from expr (log2(TPM + 0.001)) to standard log2(TPM + 1) format
expr <- log2((2**expr - 0.001) + 1)

# Step 3: Filter Low/Zero Expression Genes
# Count samples with expr >= 0.5 per transcript
count_over_0_5 <- as.numeric(apply(expr, 1, function(i) sum(i >= 0.5)))

# Identify transcripts with sufficient expression levels
keep_transcripts <- which(count_over_0_5 >= dim(expr)[2] * 0.3)

# Filter for transcripts meeting criteria
expr <- expr[keep_transcripts, ]    # Dimensions: 15,857 x 148 - including 15k protein-coding genes.

# Display
DT::datatable(expr[1:8, 1:4])

```

### Perform PCA:


```{r perform-pca}

# Transpose expr data for PCA , samples as rows and genes as columns.
# Position samples as rows and genes as columns
expr_t <- t(expr) #samples as rows and genes as columns

# Calculate PCA 
pc <- prcomp(expr_t, center = TRUE, scale. = TRUE)

# Calculate and print the percentage of variance explained by each principal component
var_res <- pc$sdev^2 / sum(pc$sdev^2) * 100
var_res <- round(var_res, 2)  # Round to 2 decimal places

print(var_res)
```

Merge PCA and Clinical Data Subset

```{r merge-data}

# Subsetting 'clin' data frame to include only 'patientid' and 'Institution' columns

clin_subset <- clin[, c("patientid", "Institution")]
clin_subsetresponse <-clin[, c("patientid", "response")]

# Merge PCA results with clinical data
pcx <- data.frame(pc$x)
pcx$patientid <- rownames(pcx)
pcx$patientid <- gsub("\\_", "-", pcx$patientid)

pcx_merge <- merge(pcx, clin_subset, by="patientid")
rownames(pcx_merge) <- pcx_merge[,1]
pcx_merge <- pcx_merge[,-1] # institution column is added

```


## Visualization


- **Bar Plot** : To visualize variance by each principal component.

The bar plot displays PC1 as the longest, explaining the most variance, while subsequent bars (PC2, PC3, etc.) become progressively shorter, indicating diminishing variance. 


```{r scree-plot}
#To find the largest PC of data

barplot(var_res, main="Bar Plot", xlab="Principal Component", ylab="Percentage of Variance ", col="skyblue")
```

### 1. PCA Results Plot

-  To visualize variance by each principal component.

PCA on gene expression data, shows that the 'Institution' factor has no significant effect on gene expression data sampling.
(no Batch effect has been detected)


```{r pca-results-plot}
# Create labels for the plot
xlab_text <- paste("PC1 (", var_res[1], "%)", sep = "")
ylab_text <- paste("PC2 (", var_res[2], "%)", sep = "")

# Plot PCA results
ggplot(pcx_merge, aes(PC1, PC2, color = Institution)) +
  theme_bw() +
  geom_point() +
  labs(x = xlab_text, y = ylab_text)

```

#### 2.Limma approach  

- Used to find the differentially expressed genes between responders and non-responders

```{r diff expression, include=TRUE, result = TRUE}

# Remove rows with missing values in response and expression data
na_response <- is.na(clin$response)

# Create filtered response data by excluding rows with response missing values
filtered_response <- clin$response[!na_response] 

# Access the expression data and exclude rows with missing response values
filtered_expression <- expr[, !na_response] #27 rows removed due to missing response values
 

# Create a design matrix using the filtered response data
design <- model.matrix(~ filtered_response)

# Fit a linear model using the filtered expression data and the design matrix
fit <- lmFit(filtered_expression, design)

# Perform eBayes analysis
fit <- eBayes(fit)

# To display the results in a table format, you can use 'datatable'
datatable(topTable(fit))

```

### 3. Volcano plot Prepration

- Preparing Data for volcano Plot 

- Convert fit to Data frame and add column gene_name symbol

```{r prepare volcano, include=TRUE, result = TRUE}

# Retrieve all results from the analysis
volcano_data <- topTable(fit, number=Inf) 

# Convert the result to a data frame
df <- as.data.frame(volcano_data)

# Subset 'gene_id' and 'gene_name' from the gene data
subset_annot <- annot_proteincoding[, c("gene_id", "gene_name")]

# Add a 'gene_id' column to the volcano_data
volcano_data$gene_id <- rownames(volcano_data)

# Merge 'volcano_data' and 'annot_proteincoding' by 'gene_id'
merge_result <- merge(volcano_data, subset_annot, by= "gene_id")

# Display the merged result in a table format
datatable(merge_result)


```

### Volcano Plot

```{r volcan plot, include=TRUE, result = TRUE}

# Volcano Plot based on P value 
EnhancedVolcano(merge_result,
    lab = merge_result$gene_name,
    x = 'logFC',
    y = 'P.Value',  
    pCutoff = 0.05,
    FCcutoff = 1.5,
    xlim = c(-5, 5),
    ylim = c(0, -log10(10e-4)),
    title= 'Volcano Plot based on P value'
    )


# Volcano Plot based on FDR 


EnhancedVolcano(merge_result,
    lab = merge_result$gene_name,
    x = 'logFC',
    y = 'adj.P.Val',
    pCutoff = 0.05,
    FCcutoff = 1.5,
    xlim = c(-5, 5),
    ylim = c(0, -log10(10e-4)),
    legendLabels=c('NS.','Log FC','adj p-value',
      'adj p-value & Log FC'),
    title= 'Volcano Plot based on FDR',)


```



### 4. Figure 1(a) of paper

- First 1.(a)_RNA cohort in different institutions.

```{r piechart , include=TRUE, result = TRUE}


clin <- data.frame(colData(mae))
clin_subset <-clin[,c("Institution","rna")]


# Counting the occurrences of each institution
institution_counts <- table(clin_subset$Institution)

# Creating the pie chart
pie_chart <- ggplot(clin_subset, aes(x = "", fill = factor(Institution))) +
  geom_bar(width = 1, stat = "count") +
  coord_polar("y", start = 0) +
  theme_void() +
  theme(legend.title = element_blank()) +
  scale_fill_manual(values = c("#999999", "#E69F00", "#56B4E9",
                               "#009E73", "#F0E442", "#0072B2",
                               "#D55E00", "#CC79A7", "#999999",
                               "#999999", "#999999", "#999999",
                               "#999999", "#999999", "#999999"))

# Printing the pie chart
print(pie_chart)

```

### 5. Figure 1(b) of paper


Figure 1(b):BOR/recist by PDL1 TPS category.


```{r bar plot, include=TRUE, result = TRUE}


# Select only the necessary columns
clin <- clin[, c("PDL1_TPS_Description", "recist", "PDL1_TPS")]

# Convert NA values to a specific level
clin$PDL1_TPS_Description <- forcats::fct_explicit_na(clin$PDL1_TPS_Description, na_level = "Unknown")
clin$PDL1_TPS_Description <- factor(clin$PDL1_TPS_Description, levels = c("<1%", "1-49%", ">50%", "Unknown"))

clin$recist <- forcats::fct_explicit_na(clin$recist, na_level = "NE")

# Count the occurrences of combinations of 'PDL1_TPS_Description' and 'recist'
# Note: Do not quote the column names
counted_data <- dplyr::count(clin, PDL1_TPS_Description, recist)

# Group by 'PDL1_TPS_Description' and calculate frequencies
grouped_data <- group_by(counted_data, PDL1_TPS_Description)
data_summary <- mutate(grouped_data, freq = n / sum(n))

# Plotting
ggplot(data_summary, aes(x = PDL1_TPS_Description, y = freq, fill = recist)) +
  geom_bar(stat = "identity") +
  scale_y_continuous(labels = scales::percent) +
  labs(x = "PDL1 TPS", y = "BOR Proportion", fill = "RECIST Category") +
  geom_text(aes(label = scales::percent(freq)), position = position_stack(vjust = 0.5)) +
  theme_minimal()


```