---
title: "QC Analysis of ICB_Ravi mae"
author: "Nasim Bondar Sahebi"
date: "2023-11-07"
output:
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```


# load libraries

```{r libraries}
library(ggplot2)
library(dplyr)
library(MultiAssayExperiment) 
library(limma)
library(EnhancedVolcano)
library(DT)
library(RColorBrewer)
library(readr)
library(ggVennDiagram)
```


# Data Loading and Preparation

**Data Overview**:

  - **Study Reference**: [PubMed ID 37024582](https://pubmed.ncbi.nlm.nih.gov/37024582/)
  - **Patient Count (RNA)**: 148
  - **Treatment Details**:
  - **PD-1/PD-L1**: 130 patients
  - **IO+chemo**: 2 patients
  - **IO+combo**: 16 patients

Load multiassay .rds file, extract clinical, expression and annotations data; prepare gene expression data for analysis.

```{r load-expression-data}
# Load your multiassay result and extract clinical data , expression data and annotation

#load mae obj
mae <- readRDS("~/BHK lab/Ravi_Testing/output/ICB_Raviii.rds")

#extract Clinical data 
clin <- data.frame(colData(mae))

#extract the expression data
expr <- assays(mae)[["expr"]]

#extracting the annotation 
annot <- data.frame(rowData(mae@ExperimentList$expr))

# Display first few rows of the dataset
DT::datatable(expr[1:8, 1:4])
```

# Principal Component Analysis (PCA)

**Objective**: Perform PCA on gene expression data to assess variation across different research centers. 

**Data Preparation**: Following the **RNA-seq differential expression**  analysis section from a paper, the analysis of differentially expressed genes involved filtering for protein-coding transcripts with a minimum expression of log2TPM ≥ 0.5 in at least 30% of samples. 

Therefore, steps for PCA-ready gene expression data include:

1. Restrict expression data to protein-coding genes.
2. Convert expression data from `log2(TPM+0.001)` to `log2(TPM+1)` for consistency.
3. Remove low/zero expression genes; Keep genes with expression at `log2TPM ≥ 0.5` across at least 30% of samples.


```{r protein-coing genes}
# Step 1: Restrict to Protein-Coding Genes
# Filter out non-protein-coding transcripts for focused analysis
annot_proteincoding <- annot[annot$gene_type == "protein_coding",] # 19158 protein coding genes.
expr <- expr[rownames(expr) %in% rownames(annot_proteincoding),]

# Step 2: Normalize Expression Data
# Convert from expr (log2(TPM + 0.001)) to standard log2(TPM + 1) format
expr <- log2((2**expr - 0.001) + 1)

# Step 3: Filter Low/Zero Expression Genes
# threshold for expr >= 0.5 per transcript
thresh <- expr >= 0.5

# Identify and keep transcripts with sufficient expression levels
keep <- rowSums(thresh) >= (0.3 * ncol(expr))
expr <- expr[keep,]  # Dimensions: 15,857 x 148 - including 15k protein-coding

# Display expr
DT::datatable(expr[1:8, 1:4])

```

## Perform PCA

- Calculate PCA
- Then Merge PCA and Clinical Data Subset


```{r perform-pca}

#1. PCA 

# Transpose expr data for PCA , samples as rows and genes as columns.
expr_t <- t(expr) #samples as rows and genes as columns

# Calculate PCA 
pc <- prcomp(expr_t, center = TRUE, scale. = TRUE)

# Calculate and print the percentage of variance explained by each principal component
var_res <- pc$sdev^2 / sum(pc$sdev^2) * 100
var_res <- round(var_res, 2)  # Round to 2 decimal places

print(var_res)

# 2. Merge PCA and Clinical Subset

# Convert PCA results to a data frame
pcx <- data.frame(pc$x)

# Add patient IDs from row names to the data frame
pcx$patientid <- rownames(pcx)


# Merge PCA data with patient IDs and institutions from clinical data
pcx_merge <- merge(pcx, clin[, c("patientid", "Institution")], by="patientid")

# Set row names to patient IDs
rownames(pcx_merge) <- pcx_merge[,1]

# Removeredundant patient ID column
pcx_merge <- pcx_merge[,-1] 

```

## Visualization


- **Bar Plot** : To visualize variance by each principal component.

The bar plot displays PC1 as the longest, explaining the most variance, while subsequent bars (PC2, PC3, etc.) become progressively shorter, indicating diminishing variance. 


```{r scree-plot}
#To find the largest PC of data

barplot(var_res, main="Bar Plot", xlab="Principal Component", ylab="Percentage of Variance ", col="skyblue")
```

## 1. PCA Results Plot

-  **Reason** :To visualize variance by each principal component.

-  **Result** :PCA on gene expression data, shows that the 'Institution' factor has no significant effect on gene expression data sampling.
(no Batch effect has been detected)


```{r pca-results-plot}
# Create labels for the plot
xlab_text <- paste("PC1 (", var_res[1], "%)", sep = "")
ylab_text <- paste("PC2 (", var_res[2], "%)", sep = "")

# Plot PCA results
ggplot(pcx_merge, aes(PC1, PC2, color = Institution)) +
  theme_bw() +
  geom_point() +
  labs(x = xlab_text, y = ylab_text)


```

## 2. Limma Approach

- This approach is used to find differentially expressed genes between responders and non-responders.

### Compare Responders and Nonresponders Using Limma Voom:

1. **Analysis with `Harmonized_Confirmed_BOR_Bin`:**
   - Apply limma voom for differential expression using the `Harmonized_Confirmed_BOR_Bin` from the clinical dataset, aiming to distinguish responders (PR/CR) from nonresponders (SD/PD). This was mentioned in paper method section **RNA-seq Differential Expression**

2. **Analysis `response`:**
   - Further differential expression analysis is conducted, utilizing the `response` classification obtained via the [`Get_Response`](https://github.com/BHKLAB-DataProcessing/ICB_Common/blob/main/code/Get_Response.R) script.
  

```{r diff expression, include=TRUE, result = TRUE}

# Filter expression data based on 'Harmonized_Confirmed_BOR_Bin' column in 'clin'
bor_filtered <- clin$Harmonized_Confirmed_BOR_Bin[!is.na(clin$Harmonized_Confirmed_BOR_Bin)]
expr_bor <- expr[, !is.na(clin$Harmonized_Confirmed_BOR_Bin)] # Remove 12 rows with missing responses

# Create design matrix and fit linear model
design_bor <- model.matrix(~ bor_filtered)
fit_bor <- lmFit(expr_bor, design_bor)

# Perform eBayes analysis and display results
fit_bor <- eBayes(fit_bor)
datatable(topTable(fit_bor))

# Repeat process for 'response' column in 'clin'
response_filtered <- clin$response[!is.na(clin$response)]
expr_response <- expr[, !is.na(clin$response)] # Remove 27 rows with missing responses

# Create design matrix and fit linear model
design_response <- model.matrix(~ response_filtered)
fit_response <- lmFit(expr_response, design_response)

# Perform eBayes analysis and display results
fit_response <- eBayes(fit_response)
datatable(topTable(fit_response))



```

### 3. Volcano plot Prepration 

- Preparing Data for volcano Plot 

- Convert fit, both `fit_bor` and `fit_response`, to Data frame and add column `gene_name` symbol 

```{r prepare volcano, include=TRUE, result = TRUE}

# Retrieve all results from the analysis and convert
volcano_data_res <- topTable(fit_bor, number=Inf)
volcano_data_bor <- topTable(fit_bor, number=Inf)

# Convert the result to a data frame
df_res <- as.data.frame(volcano_data_res)
df_bor <- as.data.frame(volcano_data_bor)

# Subset 'gene_id_no_ver' and 'gene_name' from the gene data
subset_annot <- annot_proteincoding[, c("gene_id_no_ver", "gene_name")]

# Add a 'gene_id' column to the volcano_data
volcano_data_res$gene_id_no_ver <- rownames(volcano_data_res)
volcano_data_bor$gene_id_no_ver <- rownames(volcano_data_bor)

# Merge 'volcano_data' and 'annot_proteincoding' by 'gene_id'
merge_result_res <- merge(volcano_data_res, subset_annot, by = "gene_id_no_ver")
merge_result_bor <- merge(volcano_data_bor, subset_annot, by = "gene_id_no_ver")

# Display the merged result in a table format
datatable(merge_result_res)
datatable(merge_result_bor)


```

### Volcano plot

Fig. 2 in the paper details a Volcano plot from limma voom analysis, displaying two-sided nominal P values. Genes are identified as significantly differentially expressed using cutoffs of `|log2(fold change)| > 0.5` and `P < 0.05`.



```{r volcan plot, include=TRUE, result = TRUE ,fig.width= 12 , fig.height= 10 }

# 1.Volcano Plot based on P value:

# For merge_result_res:

EnhancedVolcano(merge_result_res,
    lab = merge_result_res$gene_name,
    x = 'logFC',
    y = 'P.Value',  
    pCutoff = 0.05,
    FCcutoff = 0.5,
    xlim = c(-2, 2),
    ylim = c(0, -log10(10e-4)),
    title= 'Volcano Plot based on P value (based on BOR)'
    )

# For merge_result_res:

EnhancedVolcano(merge_result_bor,
    lab = merge_result_bor$gene_name,
    x = 'logFC',
    y = 'P.Value',
    pCutoff = 0.05,
    FCcutoff = 0.5,
    xlim = c(-2, 2),
    ylim = c(0, -log10(10e-4)),
    title= 'Volcano Plot based on P value (Based on response)'
    )


```
